### ISSUE TITLE
Uncertainty About Total Length When Using Multiple Training Datasets

### ISSUE THREAD

**Issue by dtoniolo** [PyTorchLightning/pytorch-lightning#11475]
Hello there and thanks for your awesome work.

I'm training some models for which I need to use two different `DataLoader`s. [The documentation](https://pytorch-lightning.readthedocs.io/en/stable/guides/data.html#multiple-training-dataloaders) explains nicely how to do this, but it's not clear to me on how the total length is determined, nor if there's some way of controlling the choice .

Right now in my tests the total length seems to be set equal to the one of the longest dataset. However, I would need to have the total length to be always determined by the first dataset, while the second one is either truncated or cycled through if it is smaller or larger than the first one. Is there a way to force this behaviour?

cc @borda @rohitgr7 @justusschock @awaelchli @ninginthecloud

**Comment by rohitgr7** [comment]
hey @dtoniolo !

right now this isn't possible, but I think maybe we can easily configure this by allowing users to return `CombinedLoader` instance with the mode of their choice instead of a strict dataloader from `train_dataloader`.

cc: @justusschock

**Comment by justusschock** [comment]
@rohitgr7 unfortunately this won't be possible when returning the `CombinedLoader` directly as well. The `CombinedLoader` currently only supports two modes: Using the longest one and repeating the others until this one is exhausted or using the shortest one and truncating the others. However, this cannot be controlled to always use a specific loader and change the mode based on its length in relation to the others

**Comment by rohitgr7** [comment]
> @rohitgr7 unfortunately this won't be possible when returning the CombinedLoader directly as well.

@justusschock this won't work if we enable it?
```py
def train_dataloader(self):
    dl1 = ...
    dl2 = ...
    
    if len(dl1) > len(dl2):
        return CombinedLoader([dl1, dl2], mode='max_size_cycle')
    else:
        return CombinedLoader([dl1, dl2], mode='min_cycle')
```

**Comment by justusschock** [comment]
@rohitgr7 ah okay, yeah this could work, but it definitely requires some additional logic from the user and won't work if you have 3 loaders and the loader you want to control everything is neither the longest, nor the shortest one.

**Comment by dtoniolo** [comment]
This is exactly what I was looking for and suffices for my use case, thank to both of you. I wasn't aware of the existence of `CombinedLoader`, my bad.

My issue is solved, but other developers may run in a similar problem. Should I close the issue or leave it open as a feature request for the implementation of a more sophisticated behaviour of `CombinedLoader`?

**Comment by rohitgr7** [comment]
@dtoniolo this won't work yet since returning a CombinedLoader isn't supported yet.

> @rohitgr7 ah okay, yeah this could work, but it definitely requires some additional logic from the user and won't work if you have 3 loaders and the loader you want to control everything is neither the longest, nor the shortest one.

I'd say we can at least enable this dynamic behavior for now.

**Comment by dtoniolo** [comment]
Are you sure that it is not supported? I'm just genuinely being curious, because [the docs](https://pytorch-lightning.readthedocs.io/en/latest/guides/data.html) seem to suggests it's possible. In any case, tomorrow I'll make a test as soon as possible. Also, for my use case I prefer to avoid making use of the `train_dataloader()` method and instead i create the data loader in my main script, which I then pass as an argument to the `fit()` method. I'll test this too.

**Comment by rohitgr7** [comment]
> Are you sure that it is not supported? I'm just genuinely being curious, because [the docs](https://pytorch-lightning.readthedocs.io/en/latest/guides/data.html) seem to suggests it's possible. In any case, tomorrow I'll make a test as soon as possible. Also, for my use case I prefer to avoid making use of the `train_dataloader()` method and instead i create the data loader in my main script, which I then pass as an argument to the `fit()` method. I'll test this too.

```
Evaluation dataloaders are iterated over sequentially. If you want to iterate over them in parallel, PyTorch Lightning provides a CombinedLoader object which supports collections of dataloaders such as list, tuple, or dictionary. The dataloaders can be accessed using in the same way as the provided structure
```
yes, it's only supported for val/test/predict dataloaders since for training it expects native DataLoaders and then it wraps it around `CombinedLoader`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/033dba1494a177954e8ca59bc74b1635e83b9efa/pytorch_lightning/trainer/trainer.py#L1769-L1801

even passing it to `.fit` won't work as of now.

**Comment by dtoniolo** [comment]
I'm testing it right now and the first epoch seems to run fine. As soon as it ends I'll be able to test if there are issues when an epoch after the first one is run

**Comment by ninginthecloud** [comment]
Hi, @rohitgr7  and @justusschock , this issue is exaIctly a use case that previous closed feature issue (#9046 ) can solve  Could we re-evaluate  and open the issue #9046 ? In short, #9046 support for multiple dataloaders in training and evaluation/validation/testing/prediction. It provides the user flexibility to update how they want to iterate over multiple dataloaders via `DataIterationStrategy`. It is similar to `CombinedLoader`, but it provides iteration strategy more than just 'max_size_cycle' and 'min_cycle'.
